Enhance ErrorReport Instrumentation

hxiong388 opened the pull request:
This PR primarily introduces a super simple one-liner API `Google::Cloud::ErrorReporting.report` for users to report exceptions to Stackdriver ErrorReporting. It also includes an instrumentation client that only supports error event report API and a new configuration interface designed to enhance user experience.

Example:
```ruby
require "google/cloud/error_reporting"
 
# Configure Stackdriver ErrorReporting instrumentation
Google::Cloud::ErrorReporting.configure do |config|
  config.project_id = "my-project"
  config.keyfile = "/path/to/keyfile.json"
end
 
# Insert a Rack Middleware to report unhanded exceptions 
use Google::Cloud::ErrorReporting::Middleware
 
# Or explicitly submit exceptions
begin
  fail "Boom!"
rescue => exception
  Google::Cloud::ErrorReporting.report exception
end
```

The configuration mechanism introduced here will soon be applied to all the other Stackdriver instrumentation libraries.

I also re-wrote most of the google-cloud-error_reporting documentation in response to feedback from alpha customer, where the previous ErrorReporting (and Monitoring) documentation was inconsistent with rest of the google-cloud-ruby libraries and confusing. The documentation now primarily focuses on the instrumentation workflows for best user experience. We can add GAPIC documentation once the team decides on a documentation strategy for GAPIC.

hxiong388 commented:
The `bundle exec rake ci` task is failing on me with same 98 Datastore errors on master branch. How is this happening??

blowmage commented:
The master branch has been fixed. Please rebase your branch on the current master.

quartzmo commented on google-cloud-error_reporting/docs/instrumentation.md:
````diff
@@ -1,39 +1,150 @@
 use Google::Cloud::ErrorReporting::Middleware
 ```
+
+## Report Captured Expcetions
````
s/Expcetions/Exceptions/

quartzmo commented on google-cloud-error_reporting/docs/instrumentation.md:
````diff
@@ -1,39 +1,150 @@
 ```
+
+## Report Captured Expcetions
+Captured Ruby exceptions can be directly reported to Stackdriver ErrorReporting
````
s/ErrorReporting/Error Reporting/

quartzmo commented on google-cloud-error_reporting/docs/instrumentation.md:
````diff
@@ -1,39 +1,150 @@
+end
+```
+
+The reported error event can also be customized
````
Add a period `.` or a colon `:`, and a newline after this line.

quartzmo commented on google-cloud-error_reporting/docs/instrumentation.md:
````diff
@@ -1,39 +1,150 @@
+    error_event.http_status = 502
+  end
+end
+```
````
Add a newline.

hxiong388 replied:
Should be fixed in https://github.com/GoogleCloudPlatform/google-cloud-ruby/pull/1454/commits/f7c436756604ca46d2e752c29162a94ea7b554ab

quartzmo commented on google-cloud-error_reporting/lib/google/cloud/error_reporting.rb:
```diff
@@ -0,0 +1,258 @@
+module Google
+  module Cloud
+    ##
+    # # Stackdriver ErrorReporting
```
I might remove this line, or else s/ErrorReporting/Error Reporting/

quartzmo replied:
Or: s/Stackdriver ErrorReporting/ErrorReporting/

quartzmo commented on google-cloud-error_reporting/lib/google/cloud/error_reporting.rb:
```diff
@@ -0,0 +1,258 @@
+    # Stackdriver Error Reporting counts, analyzes and aggregates the crashes in
+    # your running cloud services. The Stackdriver Error Reporting API provides:
+    # * [A simple endpoint to report errors from your application](
+    # #report-error).
```
I'm not sure this anchor link will work in the gh-pages site. :(

hxiong388 replied:
The same anchor syntax is [currently being used](https://github.com/GoogleCloudPlatform/google-cloud-ruby/blob/ced1cb4de543ce06312f13ef816a370c18cab76d/google-cloud-logging/lib/google/cloud/logging.rb#L29-L34), and it seems to be working on the github.io docs.

quartzmo replied:
If it works, great, please disregard my comment. And thanks for checking.

hxiong388 replied:
It didn't work, because of the new line. :( 

I'm fixing this anchor link.

quartzmo commented on google-cloud-error_reporting/lib/google/cloud/error_reporting.rb:
```diff
@@ -0,0 +1,258 @@
+    # ## Report Error
+    #
+    # You can easily report exceptions from your applications to Stackdriver
+    # ErrorReporting service
```
Add a period `.` or a colon `:`, and a newline after this line.

quartzmo commented on google-cloud-error_reporting/lib/google/cloud/error_reporting.rb:
```diff
@@ -0,0 +1,258 @@
+
+      ##
+      # Configure the default Google::Cloud::ErrorReporting client, allow the
+      # Google::Cloud::ErrorReporting.report public method to reuse these
```
Wrap library types in curly braces so that YARD will (hopefully) link them?

quartzmo commented on google-cloud-error_reporting/lib/google/cloud/error_reporting.rb:
```diff
@@ -0,0 +1,258 @@
+      #   * service_name: An identifier for the running service. Optional,
+      #              automatically discovered when running from Google App
+      #              Engine Flex. Otherwise default to "ruby".
+      #   * service_version: A version identifier for the running service.
```
Use backticks for these param names.

quartzmo commented on google-cloud-error_reporting/lib/google/cloud/error_reporting.rb:
```diff
@@ -0,0 +1,258 @@
+      #                      from Google App Engine Flex.
+      #
+      # Note the project_id and keyfile configuration changes won't be picked up
+      # after the first Google::Cloud::ErrorReporting.report call.
```
Curly braces for code type.

quartzmo commented on google-cloud-error_reporting/lib/google/cloud/error_reporting.rb:
```diff
@@ -0,0 +1,258 @@
+      # Note the project_id and keyfile configuration changes won't be picked up
+      # after the first Google::Cloud::ErrorReporting.report call.
+      #
+      # @example
```
We always include require and other setup in every example:

```
      #   require "google/cloud/error_reporting"
```

quartzmo commented on google-cloud-error_reporting/lib/google/cloud/error_reporting.rb:
```diff
@@ -0,0 +1,258 @@
+      # below and {Google::Cloud::ErrorReporting::ErrorEvent} class for avaiable
+      # error event fields.
+      #
+      # @example Basic usage
```
We always include require and other setup in every example:

```
      #   require "google/cloud/error_reporting"
```

hxiong388 replied:
Should be fixed in https://github.com/GoogleCloudPlatform/google-cloud-ruby/pull/1454/commits/f7c436756604ca46d2e752c29162a94ea7b554ab

quartzmo commented on google-cloud-error_reporting/lib/google/cloud/error_reporting.rb:
```diff
@@ -0,0 +1,258 @@
+      #     Google::Cloud::ErrorReporting.report exception
+      #   end
+      #
+      # @example The error event can be customized if needed
```
Again, we avoid fragments that cannot be run without missing setup, if possible.

hxiong388 replied:
Should be fixed in https://github.com/GoogleCloudPlatform/google-cloud-ruby/pull/1454/commits/f7c436756604ca46d2e752c29162a94ea7b554ab

quartzmo commented on google-cloud-error_reporting/lib/google/cloud/error_reporting/project.rb:
```diff
@@ -0,0 +1,220 @@
+      #                                             service_name: "my_app_name"
+      #   error_reporting.report error_event
+      #
+      # @see Google::Cloud::ErrorReporting.new
```
I don't think YARD's `@see` tag can be used this way, at least not in our gh-pages site. Probably better to just write:

```
See {Google::Cloud::ErrorReporting.new}.
```

quartzmo commented on google-cloud-error_reporting/lib/google/cloud/error_reporting/project.rb:
```diff
@@ -0,0 +1,220 @@
+      class Project
+        ##
+        # Find default project_id from ERROR_REPORTING_RPOJECT,
+        # GOOGLE_CLOUD_PROJECT, GCLOUD_PROJECT environment varaibles, or query
```
Backticks for these variable names.

quartzmo commented on google-cloud-error_reporting/lib/google/cloud/error_reporting/project.rb:
```diff
@@ -0,0 +1,220 @@
+
+        ##
+        # Find default service_name from ERROR_REPORTING_SERVICE,
+        # GAE_MODULE_NAME environment varaibles or just "ruby"
```
Backticks for these variable names.

s/varaibles/variables/

quartzmo commented on google-cloud-error_reporting/lib/google/cloud/error_reporting/project.rb:
```diff
@@ -0,0 +1,220 @@
+
+        ##
+        # Find default service_version from ERROR_REPORTING_VERSION or
+        # GAE_MODULE_VERSION environment varaibles
```
Backticks.

quartzmo commented on google-cloud-error_reporting/lib/google/cloud/error_reporting/project.rb:
```diff
@@ -0,0 +1,220 @@
+        ##
+        # Create a new Project instance
+        #
+        # @param [Google::Cloud::ErrorReporting::Service] service The underneath
```
Should this constructor be marked `@private` for docs? (I don't know this, just asking.)

quartzmo commented on google-cloud-error_reporting/lib/google/cloud/error_reporting/project.rb:
```diff
@@ -0,0 +1,220 @@
+        end
+
+        ##
+        # Get the name of current project_id from underneath gRPC Service object
```
s/underneath/underlying/

quartzmo commented on google-cloud-error_reporting/lib/google/cloud/error_reporting/project.rb:
```diff
@@ -0,0 +1,220 @@
+        #   uniquely identifies the user. When sending an error report, leave
+        #   this field empty if the user was not logged in. In this case the
+        #   Error Reporting system will use other data, such as remote IP
+        #   address, to distinguish affected users
```
missing final period `.`?

quartzmo commented on google-cloud-error_reporting/lib/google/cloud/error_reporting/service.rb:
```diff
@@ -0,0 +1,117 @@
+        attr_accessor :mocked_error_reporting
+
+        ##
+        # Report an ErrorEvent to Stackdriver ErrorReporting
```
Missing period `.`? Any more to say about this method?

dazuma commented on google-cloud-error_reporting/README.md:
```diff
@@ -1,82 +1,95 @@
+# google-cloud-error_reporting
+
+[Stackdriver Error Reporting](https://cloud.google.com/error-reporting/) counts,
+analyzes and aggregates the crashes in your running cloud services. A
```
s/the crashes/errors raised/

I know the current text is verbatim from cloudsite, but for Ruby I want to allow for data that is not a "crash".

dazuma commented on google-cloud-error_reporting/README.md:
```diff
@@ -1,82 +1,95 @@
+to Google Cloud services. When running on Google Cloud Platform environments, 
+the credentials will be discovered automatically. When running on other 
+environments the Service Account credentials can be specified by providing the 
+path to the JSON file, or the JSON itself, in environment variables.
```
in environment variables or configuration code

dazuma commented on google-cloud-error_reporting/docs/authentication.md:
```diff
@@ -0,0 +1,43 @@
+
+### Project and Credential Lookup
+
+Gcloud aims to make authentication as simple as possible, and provides several mechanisms to configure your system without providing **Project ID** and **Service Account Credentials** directly in code.
```
Are you talking about the GCloud SDK here? If not, use a different term.

Also, since you talk about providing Project ID and credentials _in code_ below, you might want to say "without _necessarily_ providing" or some other qualifier.

quartzmo replied:
This ("Gcloud aims to make") looks like an old version. @hxiong388 Please let me know if this was copied from somewhere in the current source or docs of the project...

hxiong388 replied:
@quartzmo Yeah, this authentication.md is yet-another-copy of the same file from all the other google-cloud-ruby libraries. We should probably fix them all together.

@dazuma I try to keep it consistent in this authenication.md, so users don't get confused when they notice a difference in these authentication instructions (one of the feedback from Monitoring customer). But I added a new section for `Instrumentation Configuration` to show additional way to provide auth info.

quartzmo replied:
@hxiong388 Makes sense. This needs to be fixed everywhere, so I will do it in a new PR once this one is merged. Thanks @dazuma for commenting on this.

dazuma commented on google-cloud-error_reporting/docs/authentication.md:
```diff
@@ -0,0 +1,43 @@
+
+### Environment Variables
+
+The **Project ID** and **Credentials JSON** can be placed in environment variables instead of declaring them directly in code. Each service has its own environment variable, allowing for different service accounts to be used for different services. The path to the **Credentials JSON** file can be stored in the environment variable, or the **Credentials JSON** itself can be stored for environments ...
```
By saying "Each service has its own environment variable, allowing for different service accounts to be used for different services" we leave out that we also have shared variables that let us set these values just once and have all services share them (which is actually what we expect the common use case to be).

Maybe consider: "You may set a single environment variable (prefixed with `GOOGLE_CLOUD_`) to share its value across all services, or you may set service-specific variables to authenticate differently per service."

hxiong388 replied:
This authentication.md maybe consistently outdated in all google-cloud-ruby libraries, but at least it's consistent. I think we should update them all in a different task.

hxiong388 commented:
@quartzmo Thanks for reviewing the PR. I have updated the code to resolve all of your comments. I've also went through all the files again and addressed these same problems.

dazuma commented on google-cloud-error_reporting/docs/instrumentation.md:
```diff
@@ -1,39 +1,165 @@
-Then the library can be configured through this set of Rails parameters in config/environments/*.rb:
+
+## Configuration
+The Stackdriver ErrorReporting instrumentation featuers should work by default 
```
Several spelling and wording issues. Try this:

The default configuration enables Stackdriver instrumentation features for most applications running on Google Cloud Platform. If you want to run on a non Google Cloud environment or you want to customize the instrumentation behavior, you can modify the Stackdriver configuration through a single interface:

dazuma commented on google-cloud-error_reporting/docs/instrumentation.md:
````diff
@@ -1,39 +1,165 @@
+end
+```
+
+The Stackdriver ErrorReporting specific configuration parameters can also be
````
Configuration specific to Stackdriver Error Reporting can also be

dazuma commented on google-cloud-error_reporting/docs/instrumentation.md:
```diff
@@ -1,39 +1,165 @@
+easily integrate with Rack based application frameworks, such as Rails and 
+Sinatra. When enabled, it automatically gathers application exceptions from 
+requests and submit the information to the Stackdriver Error Reporting service.
+ On top of that, the google-cloud-error_reporting also implements a Railtie 
```
Remove leading space on this and subsequent lines

dazuma commented on google-cloud-error_reporting/docs/instrumentation.md:
````diff
@@ -1,39 +1,165 @@
+require "google/cloud/error_reporting/rails"
+```
+
+Then the instrumentation library can also be configured through Rails 
````
s/through Rails/using the Rails/

dazuma commented on google-cloud-error_reporting/docs/instrumentation.md:
```diff
@@ -1,39 +1,165 @@
+The google-cloud-error_reporting gem provides a Rack Middleware class that can 
+easily integrate with Rack based application frameworks, such as Rails and 
+Sinatra. When enabled, it automatically gathers application exceptions from 
+requests and submit the information to the Stackdriver Error Reporting service.
```
s/submit/submits/

dazuma commented on google-cloud-error_reporting/docs/instrumentation.md:
```diff
@@ -1,39 +1,165 @@
+### Rack Integration
 
-## Rack Integration
+Other Rack base framework can also directly leverage the Middleware directly:
```
Other Rack-based frameworks

dazuma commented on google-cloud-error_reporting/docs/instrumentation.md:
````diff
@@ -1,39 +1,165 @@
 ```
+
+## Report Captured Exceptions
+Captured Ruby exceptions can be directly reported to Stackdriver Error Reporting
````
can be reported directly

dazuma commented on google-cloud-error_reporting/lib/google/cloud/error_reporting.rb:
```diff
@@ -0,0 +1,259 @@
+    # more about the options for connecting in the [Authentication
+    # Guide](https://googlecloudplatform.github.io/google-cloud-ruby/#/docs/guides/authentication).
+    #
+    # ## Report Error
```
How to report errors

dazuma commented on google-cloud-error_reporting/lib/google/cloud/error_reporting.rb:
```diff
@@ -0,0 +1,259 @@
+      # Initialize :error_reporting as a nested Configuration under
+      # Google::Cloud if haven't already
+      unless Google::Cloud.configure.option? :error_reporting
+        Google::Cloud.configure.add_options [{
```
Consider allowing add_options to take a simple hash as well, so we don't have this awkwardness here.

dazuma commented on google-cloud-error_reporting/lib/google/cloud/error_reporting/middleware.rb:
```diff
@@ -13,69 +13,90 @@
-        #   ENV["ERROR_REPORTING_SERVICE"] then "ruby". Automatically discovered
-        #   if on GAE
+        #   {Google::Cloud::ErrorReporting::Project.default_service_name}.
+        #   Automatically discovered if on Google App Engine Flex.
```
s/Flex/Flexible Environment/

dazuma commented on google-cloud-error_reporting/lib/google/cloud/error_reporting/middleware.rb:
```diff
@@ -118,19 +139,18 @@ def report_exception env, exception
 
         ##
-        # Creates a GRPC ErrorEvent based on the exception. Fill in the
+        # Creates a ErrorEvent based on the exception. Fill in the
```
s/a/an/

dazuma commented on google-cloud-error_reporting/lib/google/cloud/error_reporting/middleware.rb:
```diff
@@ -141,56 +161,74 @@ def report_exception env, exception
+        def error_event_from_exception env, exception
+          error_event = ErrorReporting::ErrorEvent.from_exception exception
+
+          error_event.user = ENV["USER"]
```
This is probably not what we want for the user field default (i.e. the unix user running the process). At worst this might be a security issue.

dazuma commented on google-cloud-error_reporting/lib/google/cloud/error_reporting/rails.rb:
```diff
@@ -13,27 +13,26 @@
+      # {Google::Cloud::ErrorReporting::Middleware} to Rack in a Rails
       # environment. It will automatically capture Exceptions from the Rails app
-      # and report them to Stackdriver Error Reporting.
+      # and report them to Stackdriver Error Reporting service.
```
...to the Stackdriver...

dazuma commented on google-cloud-error_reporting/lib/google/cloud/error_reporting/rails.rb:
```diff
@@ -131,55 +99,57 @@ def self.grpc_channel keyfile = nil
+        def self.parse_rails_config config
+          gcp_config = config.google_cloud
+          er_config = gcp_config.error_reporting
+          if gcp_config.key?(:use_error_reporting)
```
Unneeded parens

dazuma commented on google-cloud-error_reporting/lib/google/cloud/error_reporting/service.rb:
```diff
@@ -0,0 +1,118 @@
+        end
+
+        def channel
+          require "grpc"
```
Why do we require "grpc" just-in-time like this?

blowmage replied:
The require will load GRPC, and kick off all of its internal initialization. It also makes it so users cannot fork after afterwards. So we usually generally delay the require as much as possible, wherever possible.

dazuma replied:
Ah right, forgot about the fork issue.

hxiong388 commented:
@dazuma Thanks for the comments. I should've resolved them all, except the ones on authentication.md.

blowmage commented on google-cloud-error_reporting/lib/google/cloud/error_reporting/error_event.rb:
```diff
@@ -0,0 +1,312 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
```
You want 2 new lines between the copyright comments and the first bit of code. Otherwise the copyright notice will be included as the documentation for the Google constant. And every file will add a duplicate copyright to the constant's documentation.

hxiong388 replied:
Good to know. I'll watch out for this.

blowmage commented on stackdriver-core/lib/stackdriver/core/configuration.rb:
```diff
@@ -0,0 +1,125 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
```
Probably want an extra new line here.

blowmage commented on stackdriver-core/test/google/cloud/configuration_test.rb:
```diff
@@ -12,6 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+
```
The test files don't need the extra new line, since they aren't pulled into the documentation.

hxiong388 replied:
I just added it to be consistent with the other files.

blowmage replied:
:+1:

blowmage commented:
Wow! I did not mean to merge this! I put my phone in my pocket and pocket merged!

quartzmo commented:
@dazuma Anything outstanding for you on this PR? Good to merge?

dazuma commented:
It was good to merge. :)

quartzmo commented:
Oh I see, Mike left it merged. For some reason I assumed he reverted the pocket merge!
